pkg/object/httpserver/routers/radixtree/router.go

```diff
@@ -0,0 +1,681 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package radixtree
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/megaease/easegress/pkg/object/httpserver/routers"
+	"github.com/megaease/easegress/pkg/util/stringtool"
+)
+
+// art-router implementation below is a based on the original work by
+// go-chi in https://github.com/go-chi/chi/blob/master/tree.go
+// (MIT licensed). It's been heavily modified for use as a HTTP router.
+
+type (
+	nodeType uint8
+
+	// Represents leaf node in radix tree
+	muxPath struct {
+		routers.Path
+		pattern         string
+		paramKeys       []string
+		rewriteTemplate *template.Template
+	}
+
+	paths []*muxPath
+
+	// Represents node and edge in radix tree
+	node struct {
```

Comment:
Why not use and extend mature third vendors such as https://github.com/julienschmidt/httprouter. Are our requirements hard to satisfy by second-developing existing vendors?

Reply:
The code used is actually from https://github.com/go-chi/chi/blob/master/tree.go and the documentation explains